src/libsystemd/sd-netlink/nfnl-message.c

```diff
@@ -0,0 +1,310 @@
+#include <linux/nexthop.h>
+#include <stdbool.h>
+#include <unistd.h>
+
+#include "sd-netlink.h"
+
+#include "format-util.h"
+#include "netlink-internal.h"
+#include "netlink-types.h"
+#include "netlink-util.h"
+#include "socket-util.h"
+#include "util.h"
+
+static int nft_message_new(sd_netlink *nfnl, sd_netlink_message **ret, int family, uint16_t type, uint16_t flags) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        struct nfgenmsg *nfh;
+        const NLType *nl_type;
+        size_t size;
+        int r;
+
+        assert_return(nfnl, -EINVAL);
+
+        r = type_system_root_get_type(nfnl, &nl_type, NFNL_SUBSYS_NFTABLES);
+        if (r < 0)
+                return r;
+
+        if (type_get_type(nl_type) != NETLINK_TYPE_NESTED)
+                return -EINVAL;
+
+        r = message_new_empty(nfnl, &m);
+        if (r < 0)
+                return r;
+
+        size = NLMSG_SPACE(type_get_size(nl_type));
+
+        assert(size >= sizeof(struct nlmsghdr));
+        m->hdr = malloc0(size);
+        if (!m->hdr)
+                return -ENOMEM;
+
+        m->hdr->nlmsg_flags = NLM_F_REQUEST | flags;
+
+        type_get_type_system(nl_type, &m->containers[0].type_system);
+
+        r = type_system_get_type_system(m->containers[0].type_system,
+                                        &m->containers[0].type_system,
+                                        type);
+        if (r < 0)
+                return r;
+
+        m->hdr->nlmsg_len = size;
+        m->hdr->nlmsg_type = NFNL_SUBSYS_NFTABLES << 8 | type;
+
+        nfh = NLMSG_DATA(m->hdr);
+        nfh->nfgen_family = family;
+        nfh->version = NFNETLINK_V0;
+        nfh->res_id = nfnl->serial;
+
+        *ret = TAKE_PTR(m);
+
+        return 0;
+}
+
+static int sd_nfnl_message_batch(sd_netlink *nfnl, sd_netlink_message **ret, int v) {
+        struct nfgenmsg *nfh;
+        int r;
+
+        r = message_new(nfnl, ret, v);
+        if (r < 0)
+                return r;
+
+        nfh = NLMSG_DATA((*ret)->hdr);
+        nfh->nfgen_family = AF_UNSPEC;
+        nfh->version = NFNETLINK_V0;
+        nfh->res_id = NFNL_SUBSYS_NFTABLES;
+
+        return r;
+}
+
+int sd_nfnl_message_batch_begin(sd_netlink *nfnl, sd_netlink_message **ret) {
+        return sd_nfnl_message_batch(nfnl, ret, NFNL_MSG_BATCH_BEGIN);
+}
+
+int sd_nfnl_message_batch_end(sd_netlink *nfnl, sd_netlink_message **ret) {
+        return sd_nfnl_message_batch(nfnl, ret, NFNL_MSG_BATCH_END);
+}
+
+int sd_nfnl_nft_message_new_basechain(sd_netlink *nfnl, sd_netlink_message **ret,
+                                      int family,
+                                      const char *table, const char *chain,
+                                      const char *type,
+                                      uint8_t hook, int prio) {
+        sd_netlink_message *m;
+        int r;
+
+        r = nft_message_new(nfnl, ret, family, NFT_MSG_NEWCHAIN, NLM_F_CREATE | NLM_F_ACK);
```

Comment:
you initialize the return param here already even though later code in this function can fail, and leave the function half-done returning a half-setup object. The usual pattern to follow is this:

```
_cleanup(object_freep) ObjectType *o = NULL;
int r;

o = allocate_object();
if (!o)
        return -ENOMEM;

r  = object_initialize_something(o);
if (r < 0)
        return r;

r  = object_initialize_something_else(o);
if (r < 0)
        return r;

*ret = TAKE_PTR(o);
return 0;
```

i.e. allocate object into temporary, internal variable (here: `o`), set it up fully, and after everything went successfully assign it to the return param, not a moment before. And if we exit the function on some error path early, then the `_cleanup_` logic will make sure the object is freed again and the caller will never know because guarantee that on error paths the return params are untouched while on success they are unconditionally initialized.

Reply:
(this applies the same to all functions below the same: please never touch return params until everything succeeded and cannot fail anymore)

Reply:
I'm sorry, this was result of a broken squash commit, i.e. a later commit changed it to the expected pattern but the initial added commit still contained the half-initialized version. I've now amened/squashed the correct change, the line
above now works with "m" and assigns "ret" only until right before 0 is returned.